### PR TITLE
Added workflow timing

### PR DESCRIPTION
**Pull request by ccorley**
Added a single timer decorator that works for both async and sync functions.

With trace logging enabled:
```
2021-05-11 15:35:54,188 - uvicorn.error - INFO - Application startup complete.
2021-05-11 15:37:45,613 - connect.workflows.core - TRACE - Running FhirWorkflow
2021-05-11 15:37:45,613 - connect.workflows.fhir - TRACE - FhirWorkflow.validate: resource type = Patient
2021-05-11 15:37:45,707 - connect.workflows.core - TRACE - validate() id = d58d4e34-0d1a-433d-af7a-a467af898b16 elapsed time = 0.0939889s
2021-05-11 15:37:45,707 - connect.workflows.core - TRACE - transform() id = d58d4e34-0d1a-433d-af7a-a467af898b16 elapsed time = 0.0000060s
2021-05-11 15:37:45,707 - connect.workflows.core - TRACE - FhirWorkflow: incoming message type = <class 'fhir.resources.patient.Patient'>
2021-05-11 15:37:46,312 - connect.clients.kafka - TRACE - Produced record to topic FHIR-R4_PATIENT partition [0] @ offset 8804
2021-05-11 15:37:46,313 - connect.workflows.core - TRACE -  FhirWorkflow persist: stored resource location = FHIR-R4_PATIENT:0:8804
2021-05-11 15:37:46,313 - connect.workflows.core - TRACE - persist() id = d58d4e34-0d1a-433d-af7a-a467af898b16 elapsed time = 0.6056950s
2021-05-11 15:37:46,313 - connect.workflows.core - TRACE - transmit() id = d58d4e34-0d1a-433d-af7a-a467af898b16 elapsed time = 0.0000126s
2021-05-11 15:37:46,314 - connect.workflows.core - TRACE - synchronize() id = d58d4e34-0d1a-433d-af7a-a467af898b16 elapsed time = 0.0001521s
2021-05-11 15:37:46,314 - connect.workflows.core - TRACE - run() id = d58d4e34-0d1a-433d-af7a-a467af898b16 elapsed time = 0.7009690s
2021-05-11 15:37:46,315 - uvicorn.access - INFO - 127.0.0.1:54987 - "POST /fhir/Patient HTTP/1.1" 200
2021-05-11 15:37:46,319 - connect.clients.nats - TRACE - nats_sync_event_handler: received a message on EVENTS.sync 
2021-05-11 15:37:46,319 - connect.clients.nats - TRACE - nats_sync_event_handler: detected local LFH message, not storing in kafka
```

### CODE REVIEW

File: connect/support/timer.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import functools
+import time
+from connect.config import get_settings
+
+
+logger = logging.getLogger(__name__)
+
+
+def timer(func):
+    """Async decorator to print the elapsed runtime of the decorated function"""
```
Comment (dixonwhitmire) [inline connect/support/timer.py]: You may be able to switch using the `inspect` module. https://docs.python.org/3/library/inspect.html#inspect.iscoroutinefunction

The other approach, asyncio.iscoroutinefunction was used for "legacy" asyncio implementations that relied on decorators, I think.

File: connect/support/timer.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    @functools.wraps(func)
+    async def timer_wrapper(*args, **kwargs):
+        start_time = time.time()
```
Comment (dixonwhitmire) [inline connect/support/timer.py:25]: I would opt to check to see if the timing is enabled before starting/stopping timers.

**Review by dixonwhitmire — CHANGES_REQUESTED** [review]
Hi @ccorley ! This is looking good. I think you may be able to consolidate to a single decorator using the `inspect` module. Let me know what you think.
Thanks!

File: connect/support/timer.py (hunk 2 above)
Comment (ccorley) [reply connect/support/timer.py:25]: Yup, that was the other option.  done.

File: connect/support/timer.py (hunk 1 above)
Comment (ccorley) [reply connect/support/timer.py]: The net from this is that we can use one decorator, whether a method decorated with `@timer` is sync or async, but we always need to await the method due to the decorator.  So when the decorated workflow step is called from the workflow run() or from any tests, it needs to be awaited.  The code now uses one decorator.

**Review by dixonwhitmire — APPROVED** [review]

**Comment by ccorley** [comment]
Added connect/support/timer.py for timing of non-workflow functions & tested against the primary function of the /data endpoint:
```
from connect.support.timer import timer

@router.get("")
@timer
async def get_data_record(dataformat: str, partition: int, offset: int):
```
Result:
```
2021-05-12 09:07:34,317 - connect.support.timer - TRACE - get_data_record() elapsed time = 1.1635346s
```

**Pull request merged**